src/Runtime/Distributions/Automata/Automaton.Determinization.cs

```diff
@@ -205,38 +207,43 @@ bool AllDestinationsAreSame(int stateIndex)
                         return false;
                     }
 
-                    var visitedWeightedStateSet = default(Determinization.WeightedStateSet);
+                    var visited = default((Determinization.WeightedStateSet stateSet, int index));
                     var sameSetVisited =
                         destinations.Count > 1 &&
-                        stateSetsInPath.TryGetValue(destinations, out visitedWeightedStateSet);
+                        stateSetsInPath.TryGetValue(destinations, out visited);
 
-                    if (sameSetVisited && !destinations.Equals(visitedWeightedStateSet))
+                    if (sameSetVisited)
                     {
-                        // We arrived into the same state set as before, but with different weights.
-                        // This is an infinite non-converging loop. Determinization has failed
-                        return false;
-                    }
-
-                    // Add new state to the result
-                    var destinationState = builder.AddState();
-                    weightedStateSetToNewState.Add(destinations, destinationState.Index);
-                    destinationStack.Push((true, destinations));
+                        if (!destinations.EqualsWithPrecision(visited.stateSet, 1e-5))
```

Comment:
The constant for tolerance needs justification. Considering, that the goal is to check, if the difference between weights is explainable by a series of round-off errors, it seems huge. For values close to zero, it corresponds to just ~16 bits of precision out of 53. Intuitively, to allow reasonable round-off errors, we should use the relative difference and the tolerance around 2^-50.

Reply:
Also, notice that `WeightedStateSet.Equals` was already doing something similar, though, seemingly, wrong.

Reply:
It checked for the equality of 32 high bits of the corresponding weights. This doesn't seem good, because it won't consider values like 1.0 and prev(1.0) close.

Reply:
There is a reason why there are 2 implementations. I will try to add comments about that